docs/source/concepts/ledger-model/ledger-exceptions.rst

```diff
@@ -0,0 +1,178 @@
+be revised in the presence of rollback nodes. It is not enough to traverse
+the transaction tree in prefix order, because the actions under a rollback
+were rolled back.
+
+For example, a contract may be consumed by an exercise under a rollback node,
+and immediately again after the rollback node. This is allowed because the
+exercise was rolled back, and this does not represent a "double spend" of
+the same contract.
+
+We now define the "after" relation as a partial order on all the actions and
+rollback nodes of a transaction. This relation is generated by reflexivity,
+transitivity, and the following additional rules:
+
+#. If `act1` and `act2` are two actions that share a parent (e.g. two
+   top-level actions, or two actions under the same rollback node, or two
+   consequences of the same action), and `act1` precedes `act2` in the list
+   of children, then `act2` comes after `act1`.
+
+#. If `node1` is an action, and `node2` is a rollback node, and they both
+   share a parent, and `node1` precedes `node2` in the list of children,
+   then `node2` comes after `node1`.
+
+#. If `node1` and `node2` are actions or rollback nodes in the transaction
+   tree, and `node1` is a parent of `node2`, then `node2` comes after `node1`.
+
+With this modified "after" relation, the notion of consistency remains the
+same. Note that under this definition, a rollback node can be "after" its
+siblings, but never "before". The only time a rollback node comes "before"
+another node, is when that other node is a child or descendant of the
+rollback node.
+
+So in the example of a consuming exercise in a rollback node, followed by a
+consuming exercise on the same contract outside of the rollback node, neither
+exercise comes "after" the other. They are part of separate continuities,
+one of which was rolled back, and so they don't break the ledger integrity.
+
+Privacy
++++++++
+
+Rollback nodes also have an interesting effect on the notion of privacy in
+the ledger model. When projecting a transaction for a party `p`, it's
+necessary to preserve some of the rollback structure of the transaction,
+even if `p` does not have the right to observe every action under it. This
+is because we need for `p` to be able to verify that an exercise under a
+rollback (to which they are is party) is conformant, but we also need `p`
+to know that the exercise was rolled back.
+
+As such, the notion of projection for rollback nodes is defined as follows:
+When projecting a rollback node for a party `p`, we project the subtransaction
+contained in the rollback node, and then we wrap the projected subtransaction
+in a new rollback node.
+
+Subsequently, we see whether we can "normalise" the projected rollback node to
+prevent leaking information about the structure of the original transaction.
+In particular, we apply the following steps:
+
+- If the projected rollback node starts with another rollback node, for instance:
```

Comment:
Do any of these steps still apply, if the party sees the parent exercise? Looking at the first rule for example, it seems unintuitive to me that even if I see the full rollback the structure won’t be what it was originally. I know if I see the parent exercise or not, so leaking information is not an issue here afaict.

Reply:
Projection (for `p`) does not apply to consequences of an exercise that is seen by the party `p`. So this only happens for top-level rollback nodes OR rollback nodes under an exercise that was projected away (OR nested rollback nodes).

Reply:
Basically you get the original rollback structure if it wasn't projected away. Otherwise we project under the rollback and try to normalize the result.

Reply:
Makes sense, that’s what I was hoping for :+1: